core/src/main/java/org/elasticsearch/ingest/IngestMetadata.java

```diff
@@ -0,0 +1,216 @@
+        String currentFieldName = null;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            switch (token) {
+                case FIELD_NAME:
+                    currentFieldName = parser.currentName();
+                    break;
+                case START_ARRAY:
+                    if ("pipelines".equals(currentFieldName)) {
+                        Map<String, PipelineConfiguration> pipelines = new HashMap<>();
+                        while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                            if (token == XContentParser.Token.START_OBJECT) {
+                                PipelineConfiguration pipeline = new PipelineConfiguration(parser);
+                                pipelines.put(pipeline.getId(), pipeline);
+                            }
+                        }
+                        return new IngestMetadata(pipelines);
+                    }
+                    break;
+            }
+        }
+        return PROTO;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startArray("pipelines");
+        for (PipelineConfiguration pipeline : pipelines.values()) {
+            pipeline.toXContent(builder, params);
+        }
+        builder.endArray();
+        return builder;
+    }
+
+    @Override
+    public EnumSet<MetaData.XContentContext> context() {
+        return MetaData.API_AND_GATEWAY;
+    }
+
+    /**
+     * Encapsulates a pipeline's id and configuration as a map of maps
+     */
+    // We can't use Pipeline class directly in cluster state, because we don't have the processor factories around when
+    // IngestMetadata is registered as custom metadata.
+    public final static class PipelineConfiguration implements ToXContent {
+
+        private final String id;
+        private final BytesReference config;
+
+        public PipelineConfiguration(StreamInput input) throws IOException {
+            id = input.readString();
+            config = input.readBytesReference();
+        }
+
+        PipelineConfiguration(XContentParser parser) throws IOException {
+            String id = null;
+            BytesReference config = null;
+
+            XContentParser.Token token;
+            String currentFieldName = null;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                switch (token) {
+                    case FIELD_NAME:
+                        currentFieldName = parser.currentName();
+                        break;
+                    case VALUE_STRING:
+                        if ("id".equals(currentFieldName)) {
+                            id = parser.text();
+                        }
+                        break;
+                    case START_OBJECT:
+                        config = XContentFactory.jsonBuilder().startObject().value(parser.map()).endObject().bytes();
```

Comment:
nit: wondering if we can avoid parsing the map and rewriting it directly in json format. not sure there's a better way to do this.